### PR TITLE
TapeAgents export for experiment results

### PR DESCRIPTION
**Pull request by ollmer**
This PR introduces a simple compatibility layer for exporting traces of the BrowserGym experiments in a format compatible with the Tapes of the [TapeAgents framework](https://github.com/ServiceNow/TapeAgents).
What does it do:
- convert a single trace of the experiment into a single Tape
- decomposes each StepInfo step of the experiment into 3 steps of the tape agents: observation, thought, and action.
- attempts to preserve as much metadata as possible for the step and for the whole trace.
- parses action in the simplest possible way as a string with a single Python function call
- if the `think` field is present in the agent info and not empty - it creates the thought step of the tape from it.
- method to save the tape in json format

The MVP is ready but not yet tested.

Next steps in this PR: 
- test conversion over the real traces
- collect feedback from the BrowserGym team
- test tape loading on the TapeAgents side and rendering it with the tape browser

cc @recursix.

### CODE REVIEW

File: browsergym/experiments/src/browsergym/experiments/loop.py [hunk 1]
```diff
@@ -657,6 +658,75 @@ def summary_info(self) -> dict:
...
+        Return experiment trace in a format compatible with the TapeAgents framework
+        """
+        tape_steps = []
+        for step_info in self.steps_info:
+            # extract observation step
+            if step_info.obs is not None:
+                screenshot: str = ""
+                screenshot_som: str = ""
+                if "screenshot" in step_info.obs:
+                    screenshot = f"{self.exp_dir}/screenshot_step_{step_info.step}.png"
```
Comment (recursix) [inline browsergym/experiments/src/browsergym/experiments/loop.py:672]: leverage pathlib

```suggestion
                    screenshot = {self.exp_dir} / f"screenshot_step_{step_info.step}.png"
```

File: browsergym/experiments/src/browsergym/experiments/loop.py [hunk 2]
```diff
@@ -657,6 +658,75 @@ def summary_info(self) -> dict:
...
+                    step_info.obs.pop("screenshot")
```
Comment (recursix) [inline browsergym/experiments/src/browsergym/experiments/loop.py:673]: this is also popping in the stored self.steps_info, which could lead to unexpected errors if someon tries to use results after calling get_tape

File: browsergym/experiments/src/browsergym/experiments/loop.py [hunk 3]
```diff
@@ -657,6 +658,75 @@ def summary_info(self) -> dict:
...
+                    metadata={"step": step_info.step},
+                    obs=step_info.obs,
+                    screenshot=screenshot,
+                    screenshot_som=screenshot_som,
+                )
+                tape_steps.append(observation)
+
+            # extract thought step
+            think = ""
+            if isinstance(step_info.agent_info, AgentInfo):
```
Comment (recursix) [inline browsergym/experiments/src/browsergym/experiments/loop.py:687]: AgentInfo also behaves like a dict, so you could simplify by just doing .get

File: browsergym/experiments/src/browsergym/experiments/loop.py [hunk 4]
```diff
@@ -657,6 +658,75 @@ def summary_info(self) -> dict:
...
+                    step_info.obs.pop("screenshot_som")
+                observation = Observation(
+                    metadata={"step": step_info.step},
+                    obs=step_info.obs,
+                    screenshot=screenshot,
+                    screenshot_som=screenshot_som,
+                )
+                tape_steps.append(observation)
+
+            # extract thought step
```
Comment (recursix) [inline browsergym/experiments/src/browsergym/experiments/loop.py:685]: We should probably look if there is a tape_sgement first, like we discussed. Or should we wait until it's implemented?

Comment (recursix) [reply browsergym/experiments/src/browsergym/experiments/loop.py:685]: Could a tape_segment simply be a Tape that can be appened

Comment (ollmer) [reply browsergym/experiments/src/browsergym/experiments/loop.py:685]: We could merge this implementation to support existing traces and work on the segments separately. I think that segments would be either Tape or list[Step], cause trace metadata is not really needed in a segment

Comment (ollmer) [reply browsergym/experiments/src/browsergym/experiments/loop.py]: I've added support for tape_segment dicts stored in the `other_info` dict of the agent info.

**Comment by ollmer** [comment]
I've simplified this PR down to <100 lines of changes in a single file and added support of tape_segments inside the agent_info, which could be produced by TapeAgent in the AgentLab.

File: browsergym/experiments/src/browsergym/experiments/loop.py [hunk 5]
```diff
@@ -657,6 +657,94 @@ def summary_info(self) -> dict:
...
+        steps = []
+        for step_info in self.steps_info:
+            if "tape_segment" in step_info.agent_info["extra_info"]:
+                tape_segment = step_info.agent_info["extra_info"]["tape_segment"]
+            else:
+                tape_segment = self._create_tape_segment(step_info)
+            steps += tape_segment
+        metadata = dict(
+            id=str(uuid.uuid4()),
+            author="browsergym_agent",
```
Comment (gasse) [inline browsergym/experiments/src/browsergym/experiments/loop.py:678]: Maybe the agent name could be re-used here?
```suggestion
            author=f"browsergym_agent_[{agent_name}]",
```

File: browsergym/experiments/src/browsergym/experiments/loop.py [hunk 6]
```diff
@@ -657,6 +657,94 @@ def summary_info(self) -> dict:
...
+        )
+        return dict(steps=steps, metadata=metadata)
+
+    def _create_tape_segment(self, step_info: StepInfo) -> list[dict]:
+        tape_segment = []
+        # extract observation step
+        if step_info.obs is not None:
+            screenshot: str = ""
+            screenshot_som: str = ""
+            obs_dict = step_info.obs.copy()
```
Comment (gasse) [inline browsergym/experiments/src/browsergym/experiments/loop.py:689]: Should this be a deepcopy?
Drawback: it's costly
Pro: tapes can safely be altered afterwards

File: browsergym/experiments/src/browsergym/experiments/loop.py [hunk 7]
```diff
@@ -657,6 +657,94 @@ def summary_info(self) -> dict:
...
+        # extract thought step
+        think = step_info.agent_info.get("think", "")
+        if think:
+            tape_segment.append(
+                dict(kind="browsergym_thought", metadata={"step": step_info.step}, text=think)
+            )
+
+        # extract action step
+        action_str = step_info.action
+        # TODO: this is a naive way to split the arguments, make proper parsing later
```
Comment (gasse) [inline browsergym/experiments/src/browsergym/experiments/loop.py:715]: We have a parser in browsergym to extract function calls from raw text actions, you could just reuse it.

There is one caveat though, actions in browsergym can be arbitrary strings (python code, function calls, json). Even with our `function call` action set, agents can be given the freedom to execute several function calls in the same action (e.g. `type("a15", "Hello") click("a16")`).

Can TapeAgent handle arbitrary actions? Do you need the `name` and `arguments` key?

**Review by gasse — COMMENTED** [review]
Just made a few comments. Feel free to ping me for a chat @ollmer

File: browsergym/experiments/src/browsergym/experiments/loop.py (hunk 6 above)
Comment (ollmer) [reply browsergym/experiments/src/browsergym/experiments/loop.py:689]: Yep, it's a good idea. I assumed earlier that the observation fields are mostly primitive types and will be copied anyway, but it's safer to make a deep copy. As the export should rarely be called, complexity is not an issue here. I'll update it, thanks!

File: browsergym/experiments/src/browsergym/experiments/loop.py (hunk 7 above)
Comment (gasse) [reply browsergym/experiments/src/browsergym/experiments/loop.py:716]: You'll find an example here:
https://github.com/ServiceNow/BrowserGym/blob/80e063bca66e766c3ec3f3ea2913127dc1bb4b69/tests/core/test_actions_highlevel.py#L102

Comment (gasse) [reply browsergym/experiments/src/browsergym/experiments/loop.py:716]: The parser itself can be imported from here
https://github.com/ServiceNow/BrowserGym/blob/80e063bca66e766c3ec3f3ea2913127dc1bb4b69/browsergym/core/src/browsergym/core/action/parsers.py#L85

Comment (ollmer) [reply browsergym/experiments/src/browsergym/experiments/loop.py]: I've implemented it, please take a look when you have time.

**Comment by ollmer** [comment]
Thank you!

**Pull request merged**